text/0000-dyn-type.md

```diff
@@ -0,0 +1,2261 @@
+and user just need to provide
+
+```rust ,ignore
+unsafe impl<T> Derived for Mat<T> {
+    unsafe fn reduced_metadata(val: *const Self) -> usize { // the metadata of [T] is a usize, the length.
+        ptr::metadata(val).len()
+    }
+}
+```
+
+The problem here is that we cannot forward `align_of_meta` to `reduced_meta`
+since we don’t have `self` in the first place. This forces us to either
+
+1. make `reduced_meta` take only `Self::Metadata`, or
+2. make `align_of_meta` not rely on reduction i.e. the alignment is known at compile time.
+
+We call the first kind of DST a *regular DST*. This covers slice, trait object, matrix, `OsStr`, and bit slice.
+These types carry a nonzero metadata, and the reduction (and size) can be entirely derived from the metadata alone.
+
+```rust ,ignore
+trait Regular: Derived {
+    fn reduce_with_metadata(meta: Self::Metadata) -> Self::Reduced::Metadata;
+}
+```
+
+The second kind of DST are mainly thin DSTs like `CStr`, Pascal strings and length-prefixed arrays.
+These types has no metadata and thus the pointer is thin and can be used in FFI.
+The size can only be obtained by parsing the memory content itself, and thus we call these *inline DST*,
+as in the metadata are stored “inline”.
+
+```rust ,ignore
+unsafe trait InlineDst: Derived + Aligned + Thin {
+    unsafe fn reduce_with_ptr(ptr: *const ()) -> Self::Reduced::Metadata;
+}
+```
+
+Only trait objects and foreign types have no compile-time alignment. To distinguish between these and other DSTs,
+we’d like to introduce a new trait, `Aligned`, which marks the type as having a compile-time alignment.
+This also has a nice side effect of relaxing the bounds of `core::mem::align_of`.
+
+Since we want to distinguish between regular and inline DSTs, and modify the syntax for declaring a custom inline DST.
+
+```rust
+dyn type Mat<T>(regular: MatMeta) = [T];
+//              ^~~~~~~~
+
+dyn type CStr(inline) = [c_char];
+//            ^~~~~~
+```
+
+### Aliasing
+
+Representing custom DSTs through reduction allows the compiler to manipulate them as if normal Rust types,
+but there is a potential hazard regarding borrow-checking.
+
+Let’s revisit matrix. A matrix’s metadata consists of three numbers, the width, height and stride.
+The stride can be larger than the width when the rectangular slice does not include all columns.
+
+```
++----+----+====+====+----+  highlighted region:
+|  0 |  1 [  2 |  3 ]  4 |      width = 2
++----+----+====+====+----+      height = 2
+|  5 |  6 [  7 |  8 ]  9 |      stride = 5
++----+----+====+====+----+
+| 10 | 11 | 12 | 13 | 14 |
++----+----+----+----+----+
+```
+
+In the linear representation as a `[T]`, it will contain `[2, 3, 4, 5, 6, 7, 8]`.
+Note that the irrelevant entries `4, 5, 6` are included.
+This means providing mutable access to the reduced type is unsafe.
+Ideally we should represent this memory as `∃w,h,s: [([T; w], [Opaque<T>; s-w]); h]`,
```

Comment:
Is this a sigma type? not super clear what the notation entails.
I think this means: `dyn<w: usize, h: usize, s: usize> [([T; w], [Opaque<T>; s - w]); h]`.